src/xwayland/xwm/mod.rs

```diff
@@ -1284,6 +1313,75 @@ impl X11Wm {
 
         Ok(())
     }
+
+    /// Gets the current primary output as advertised by xrandr
+    pub fn get_randr_primary_output(&self) -> Result<Option<String>, ReplyError> {
+        let current_primary = self
+            .conn
+            .randr_get_output_primary(self.screen.root)?
+            .reply()?
+            .output;
+        if current_primary == x11rb::NONE {
+            return Ok(None);
+        }
+
+        let res = self
+            .conn
+            .randr_get_screen_resources_current(self.screen.root)?
+            .reply()?;
+        let info = self
+            .conn
+            .randr_get_output_info(current_primary, res.config_timestamp)?
+            .reply()?;
+        Ok(Some(
+            std::str::from_utf8(&info.name)
+                .expect("X11 protocol violation, output name not utf-8")
```

Comment:
I don't know how likely this is, but a malicious client could bring the compositor down with this expect. Is there an alternative like ignoring the output in this case?

Reply:
I honestly think this is a non-issue, as this value comes from the xorg server and the protocol explicitly states that this string is utf-8.